### ISSUE TITLE
[PATCH] move functions marked as mathomed in embed.fnc to mathoms.c

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14443]
Migrated from [rt.perl.org#123671](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123671) (status was 'resolved')

Searchable as RT123671$

**Comment by p5pRT** [comment]
### From @bulk88

#### Created by @bulk88



See attached patch\. Related to 
http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2015/01/msg225065\.html 
I created this patch and forgot about it \(29 commits ahead of blead ATM\) 
in my repo before 
http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2015/01/msg225065\.html 
was posted\.

<details><summary>Perl Info</summary>

```

Flags:
               category=core
               severity=low

Site configuration information for perl 5.21.4:

Configured by Owner at Thu Sep 18 12:08:58 2014.

Summary of my perl5 (revision 5 version 21 subversion 4) configuration:
             Derived from: 7d2b2edb94ab56333b9049a3e26d15ea18445512
             Ancestor: 19be3be6968e2337bcdfe480693fff795ecd1304
             Platform:
               osname=MSWin32, osvers=5.1, archname=MSWin32-x86-multi-thread
               uname=''
               config_args='undef'
               hint=recommended, useposix=true, d_sigaction=undef
               useithreads=define, usemultiplicity=define
               use64bitint=undef, use64bitall=undef, uselongdouble=undef
               usemymalloc=n, bincompat5005=undef
             Compiler:
               cc='cl', ccflags ='-nologo -GF -W3 -O1 -MD -Zi -DNDEBUG
-DWIN32
-D_CONSOLE -DNO_STRICT  -DPERL_TEXTMODE_SCRIPTS
-DPERL_HASH_FUNC_ONE_AT_A_TIME -DPERL_IMPLICIT_CONTEXT
-DPERL_IMPLICIT_SYS -DUSE_PERLIO -D_USE_32BIT_TIME_T',
               optimize='-O1 -MD -Zi -DNDEBUG',
               cppflags='-DWIN32'
               ccversion='12.00.8168', gccversion='', gccosandvers=''
               intsize=4, longsize=4, ptrsize=4, doublesize=8,
byteorder=1234
               d_longlong=undef, longlongsize=8, d_longdbl=define,
longdblsize=8,
longdblkind=0
               ivtype='long', ivsize=4, nvtype='double', nvsize=8,
Off_t='__int64',
lseeksize=8
               alignbytes=8, prototype=define
             Linker and Libraries:
               ld='link', ldflags ='-nologo -nodefaultlib -debug
-opt:ref,icf
-libpath:"c:\perl521\lib\CORE"  -machine:x86'
               libpth=C:\PROGRA~1\MIAF9D~1\VC98\lib
               libs=oldnames.lib kernel32.lib user32.lib gdi32.lib
winspool.lib
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib
netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib
odbc32.lib odbccp32.lib comctl32.lib msvcrt.lib
               perllibs=oldnames.lib kernel32.lib user32.lib gdi32.lib
winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib
oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib
version.lib odbc32.lib odbccp32.lib comctl32.lib msvcrt.lib
               libc=msvcrt.lib, so=dll, useshrplib=true, libperl=perl521.lib
               gnulibc_version=''
             Dynamic Linking:
               dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
               cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug
-opt:ref,icf  -libpath:"c:\perl521\lib\CORE"  -machine:x86'

Locally applied patches:
               uncommitted-changes
               a0fe7a7e75de29e59f1da0d6822dc06e5be658fe
               a261faffee83d0145642ab5d1d046c9f813bc497
               6506ab86ad1602a9ca720fcd30446dce1461d23d
               7d2b2edb94ab56333b9049a3e26d15ea18445512


@INC for perl 5.21.4:
               lib
               C:/perl521/srcnew/lib
               .


Environment for perl 5.21.4:
               HOME (unset)
               LANG (unset)
               LANGUAGE (unset)
               LD_LIBRARY_PATH (unset)
               LOGDIR (unset)
               PATH=
               PERL_BADLANG (unset)
               PERL_JSON_BACKEND=Cpanel::JSON::XS
               PERL_YAML_BACKEND=YAML
               SHELL (unset)





















```
</details>

**Comment by p5pRT** [comment]
### From @bulk88

<details><summary>0001-move-functions-marked-as-mathomed-in-embed.fnc-to-ma.patch</summary>

```diff
From 211c9208900536ccdd9ea53c042465f3b4dba5e5 Mon Sep 17 00:00:00 2001
From: Daniel Dragan <bulk88@hotmail.com>
Date: Tue, 20 Jan 2015 07:51:13 -0500
Subject: [PATCH] move functions marked as mathomed in embed.fnc to mathoms.c

Ever since commit 075eb5c9b6 mathom functions must be in mathoms.c for
their symbols to be skipped in makedef.pl on Win32 Perl. If a function is
marked 'b' in embed.fnc, regen.pl does NOT add its prototype to proto.h
(it is commented out). Without the proto.h entry, EXTERN_C will be missing
and a -DNO_MATHOMS + C++ Win32 Perl build will not link, since the C
function will have a mangled name and the symbol will not be found for
creating the perl linking library.

Also add EXTERN_C to Win32CORE, the init_Win32CORE symbol is special cased
for exporting in makedef.pl.

Perl_is_utf8_char_buf was marked as 'b' in commit 3cedd9d930
Perl_sv_copypv was marked as 'b' in commit 4bac9ae47b
---
 ext/Win32CORE/Win32CORE.c |    2 ++
 mathoms.c                 |   24 ++++++++++++++++++++++++
 sv.c                      |    8 --------
 utf8.c                    |   16 ----------------
 4 files changed, 26 insertions(+), 24 deletions(-)

diff --git a/ext/Win32CORE/Win32CORE.c b/ext/Win32CORE/Win32CORE.c
index 64104d3..524cf3f 100644
--- a/ext/Win32CORE/Win32CORE.c
+++ b/ext/Win32CORE/Win32CORE.c
@@ -47,6 +47,8 @@ XS_EXTERNAL(boot_Win32CORE)
      * should never be called though, as Win32CORE.pm doesn't use DynaLoader.
      */
 }
+
+EXTERN_C
 #if !defined(__CYGWIN__) || defined(USEIMPORTLIB)
 __declspec(dllexport)
 #endif
diff --git a/mathoms.c b/mathoms.c
index 96b3ac3..cd2da84 100644
--- a/mathoms.c
+++ b/mathoms.c
@@ -1217,6 +1217,14 @@ Perl_sv_mortalcopy(pTHX_ SV *const oldstr)
     return Perl_sv_mortalcopy_flags(aTHX_ oldstr, SV_GMAGIC);
 }
 
+void
+Perl_sv_copypv(pTHX_ SV *const dsv, SV *const ssv)
+{
+    PERL_ARGS_ASSERT_SV_COPYPV;
+
+    sv_copypv_flags(dsv, ssv, 0);
+}
+
 UV      /* Made into a function, so can be deprecated */
 NATIVE_TO_NEED(const UV enc, const UV ch)
 {
@@ -1695,6 +1703,22 @@ Perl_is_utf8_char(const U8 *s)
     return isUTF8_CHAR(s, s + UTF8SKIP(s));
 }
 
+/*
+=for apidoc is_utf8_char_buf
+
+This is identical to the macro L</isUTF8_CHAR>.
+
+=cut */
+
+STRLEN
+Perl_is_utf8_char_buf(const U8 *buf, const U8* buf_end)
+{
+
+    PERL_ARGS_ASSERT_IS_UTF8_CHAR_BUF;
+
+    return isUTF8_CHAR(buf, buf_end);
+}
+
 /* DEPRECATED!
  * Like L</utf8_to_uvuni_buf>(), but should only be called when it is known that
  * there are no malformations in the input UTF-8 string C<s>.  Surrogates,
diff --git a/sv.c b/sv.c
index d75bbe8..764c367 100644
--- a/sv.c
+++ b/sv.c
@@ -3307,14 +3307,6 @@ include SV_GMAGIC.
 */
 
 void
-Perl_sv_copypv(pTHX_ SV *const dsv, SV *const ssv)
-{
-    PERL_ARGS_ASSERT_SV_COPYPV;
-
-    sv_copypv_flags(dsv, ssv, 0);
-}
-
-void
 Perl_sv_copypv_flags(pTHX_ SV *const dsv, SV *const ssv, const I32 flags)
 {
     STRLEN len;
diff --git a/utf8.c b/utf8.c
index 8182e63..bf5a36e 100644
--- a/utf8.c
+++ b/utf8.c
@@ -319,22 +319,6 @@ Perl_uvchr_to_utf8_flags(pTHX_ U8 *d, UV uv, UV flags)
 }
 
 /*
-=for apidoc is_utf8_char_buf
-
-This is identical to the macro L</isUTF8_CHAR>.
-
-=cut */
-
-STRLEN
-Perl_is_utf8_char_buf(const U8 *buf, const U8* buf_end)
-{
-
-    PERL_ARGS_ASSERT_IS_UTF8_CHAR_BUF;
-
-    return isUTF8_CHAR(buf, buf_end);
-}
-
-/*
 =for apidoc is_utf8_string
 
 Returns true if the first C<len> bytes of string C<s> form a valid
-- 
1.7.9.msysgit.0


```
</details>

**Comment by p5pRT** [comment]
### From @steve-m-hay

On Sat Jan 24 21&#8203;:23&#8203;:13 2015\, bulk88 wrote&#8203;:
> See attached patch\. Related to 
> http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2015/01/msg225065\.html 
> I created this patch and forgot about it \(29 commits ahead of blead ATM\) 
> in my repo before 
> http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2015/01/msg225065\.html 
> was posted\.
> 

Excellent\! That fixes my C\+\+ build with current blead plus
the Win32 and Win32API\-File fixes that I GutHubbed yesterday\.

Two comments&#8203;:

\- The patch moves the documentation for Perl\_is\_utf8\_char\_buf from utf8\.c to mathoms\.c\, but leaves the documentation for Perl\_sv\_copypv in sv\.c\. Was that intentional?

\- Both functions are added to mathoms\.c below the line which says "Functions after here were made mathoms post 5\.10\.0 but pre 5\.8\.9"\, but that is not true of either one of them\. I haven't investigated whether that comment is still true otherwise\. Possibly it's long since become obsolete and should simply be removed\, but if it is still true and serves some purpose \(which isn't clear to me either\) then these new additions to mathoms\.c should be inserted higher up\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From Eirik-Berg.Hanssen@allverden.no

On Sun\, Jan 25\, 2015 at 2&#8203;:19 PM\, Steve Hay via RT \<perlbug\-followup@&#8203;perl\.org
> wrote&#8203;:

> post 5\.10\.0 but pre 5\.8\.9


&nbsp;  \.\.\. say what?


Eirik

**Comment by p5pRT** [comment]
### From @steve-m-hay

On 25 January 2015 at 13&#8203;:39\, Eirik Berg Hanssen
\<Eirik\-Berg\.Hanssen@&#8203;allverden\.no> wrote&#8203;:
>
> On Sun\, Jan 25\, 2015 at 2&#8203;:19 PM\, Steve Hay via RT
> \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>>
>> post 5\.10\.0 but pre 5\.8\.9
>
>
>   \.\.\. say what?
>

5\.10\.0 came out \(18 Dec 2007\) before 5\.8\.9 \(14 Dec 2008\)\.

Numerical order isn't always chronological order by any means\, thanks
to different maint sreams being concurrently active\.

**Comment by p5pRT** [comment]
### From Eirik-Berg.Hanssen@allverden.no

On Sun\, Jan 25\, 2015 at 2&#8203;:48 PM\, Steve Hay \<steve\.m\.hay@&#8203;googlemail\.com>
wrote&#8203;:

> On 25 January 2015 at 13&#8203;:39\, Eirik Berg Hanssen
> \<Eirik\-Berg\.Hanssen@&#8203;allverden\.no> wrote&#8203;:
> >
> > On Sun\, Jan 25\, 2015 at 2&#8203;:19 PM\, Steve Hay via RT
> > \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> >>
> >> post 5\.10\.0 but pre 5\.8\.9
> >
> >
> >   \.\.\. say what?
> >
>
> 5\.10\.0 came out \(18 Dec 2007\) before 5\.8\.9 \(14 Dec 2008\)\.
>
> Numerical order isn't always chronological order by any means\, thanks
> to different maint sreams being concurrently active\.
>

&nbsp;  Ah\, thanks\.  \(I should have realized chronological order was intended\.
Sorry\.\)


Eirik

**Comment by p5pRT** [comment]
### From @bulk88

On Sun Jan 25 05&#8203;:19&#8203;:01 2015\, shay wrote&#8203;: 
> Two comments&#8203;:
> 
> \- The patch moves the documentation for Perl\_is\_utf8\_char\_buf from
> utf8\.c to mathoms\.c\, but leaves the documentation for Perl\_sv\_copypv
> in sv\.c\. Was that intentional?

Yes\. sv\_copypv the macro is not mathomed\, only Perl\_sv\_copypv the C function is\. sv\_copypv the macro's docs should not appear in section "Functions in file mathoms\.c" http&#8203;://perldoc\.perl\.org/perlapi\.html \. I moved Perl\_is\_utf8\_char\_buf's docs since the docs just say "This is identical to the macro L\</isUTF8\_CHAR>\."\. That sentence probably means it has been "removed from use in new code\."\, since "we aren't documenting what it does anymore\, if you code still uses it\, that is fine\, but we wont tell you what it does so you can't use it in new code without reverse engineering \(looking at src code/clicking the link\)"\.

> 
> \- Both functions are added to mathoms\.c below the line which says
> "Functions after here were made mathoms post 5\.10\.0 but pre 5\.8\.9"\,
> but that is not true of either one of them\. I haven't investigated
> whether that comment is still true otherwise\. Possibly it's long since
> become obsolete and should simply be removed\, but if it is still true
> and serves some purpose \(which isn't clear to me either\) then these
> new additions to mathoms\.c should be inserted higher up\.

5\.21\.8 is after 5\.10\.0 right :\-\) ?

http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/09d7a3ba918e6e685992034f67fd84290a3df894 from 2014 put a function at line ~1773 \, the "/\* Functions after here were made mathoms post 5\.10\.0 but pre 5\.8\.9 \*/" comment is at line 1022\.  Commit http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/0d7d409d8d92b77ed7de5b74ab047eced86edfc3 from 2010 \(2 years after 5\.8\.9\) put a function at line 1519 \. Atleast FC and davem are ignoring or didn't see that comment\.

That comment is probably dead since the 2 track 5\.8 and 5\.9 development model is long dead\.
\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @steve-m-hay

On Sun Jan 25 15&#8203;:54&#8203;:39 2015\, bulk88 wrote&#8203;:
> On Sun Jan 25 05&#8203;:19&#8203;:01 2015\, shay wrote&#8203;:
> > Two comments&#8203;:
> >
> > \- The patch moves the documentation for Perl\_is\_utf8\_char\_buf from
> > utf8\.c to mathoms\.c\, but leaves the documentation for Perl\_sv\_copypv
> > in sv\.c\. Was that intentional?
> 
> Yes\. sv\_copypv the macro is not mathomed\, only Perl\_sv\_copypv the C
> function is\. sv\_copypv the macro's docs should not appear in section
> "Functions in file mathoms\.c" http&#8203;://perldoc\.perl\.org/perlapi\.html \. I
> moved Perl\_is\_utf8\_char\_buf's docs since the docs just say "This is
> identical to the macro L\</isUTF8\_CHAR>\."\. That sentence probably means
> it has been "removed from use in new code\."\, since "we aren't
> documenting what it does anymore\, if you code still uses it\, that is
> fine\, but we wont tell you what it does so you can't use it in new
> code without reverse engineering \(looking at src code/clicking the
> link\)"\.
> 
> >
> > \- Both functions are added to mathoms\.c below the line which says
> > "Functions after here were made mathoms post 5\.10\.0 but pre 5\.8\.9"\,
> > but that is not true of either one of them\. I haven't investigated
> > whether that comment is still true otherwise\. Possibly it's long
> > since
> > become obsolete and should simply be removed\, but if it is still true
> > and serves some purpose \(which isn't clear to me either\) then these
> > new additions to mathoms\.c should be inserted higher up\.
> 
> 5\.21\.8 is after 5\.10\.0 right :\-\) ?
> 
> http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/09d7a3ba918e6e685992034f67fd84290a3df894
> from 2014 put a function at line ~1773 \, the "/\* Functions after here
> were made mathoms post 5\.10\.0 but pre 5\.8\.9 \*/" comment is at line
> 1022\.  Commit
> http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/0d7d409d8d92b77ed7de5b74ab047eced86edfc3
> from 2010 \(2 years after 5\.8\.9\) put a function at line 1519 \. Atleast
> FC and davem are ignoring or didn't see that comment\.
> 
> That comment is probably dead since the 2 track 5\.8 and 5\.9
> development model is long dead\.

Ok\, thanks\. Patch now applied in commit e4524c4c2a and out\-dated comment removed by commit 97fc4b2276\.

**Comment by p5pRT** [comment]
@steve-m-hay - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks for submitting this ticket

The issue should be resolved with the release today of Perl v5\.22\, available at http&#8203;://www\.perl\.org/get\.html
If you find that the problem persists\, feel free to reopen this ticket

\-\- 
Karl Williamson for the Perl 5 porters team

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
